enhancements/node-latency-profile/node-latency-profiles.md

```diff
@@ -0,0 +1,276 @@
+---
+title: kubelet-cm-communication-profiles
+authors:
+  - "@harche"
+reviewers:
+  - "@rphillips, @sttts, @soltysh"
+approvers:
+  - "@rphillips, @sttts, @soltysh"
+creation-date: 2021-09-28
+last-updated: 2021-09-28
+status: implementable
+see-also:
+  - https://github.com/kubernetes-sigs/kubespray/blob/master/docs/kubernetes-reliability.md
+replaces:
+superseded-by:
+---
+
+# NodeLatencyProfile Profiles
+
+## Release Signoff Checklist
+
+- [x] Enhancement is `implementable`
+- [x] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+
+## Summary
+
+To make sure Openshift cluster keeps running optimally under various circumstances where network latency, API Server latency, etcd latency, control plane load latency, etc. may not always be at their best (e.g. Edge cases), we can tweak the frequency of the status updates done by the `Kubelet` and the corresponding reaction times of the `Controller Manager` as well as `Kube API Server`.
```

Comment:
My assumption was that this enhancement is just about worker nodes. Now I read about API and etcd latency. This puts this enhancement on a completely different level. The question becomes: do we want to support use-cases where etcd connectivity is slow? I am not convinced. This is a business and support question, not only technical.

cc @mfojtik

Reply:
can we get a clarification on what is meant with "etcd latency" and "API server latency" and "control plane latency" ? the summary sounds a little vague and generic and open a pandora box to allow unbound customization of control plane performance flags :-) 

enhancement down does not mention etcd at all and the kubespray does not mention API server flags.

Reply:
References to etcd and other latencies are borrowed from kubespray [documentation](https://github.com/kubernetes-sigs/kubespray/blob/master/docs/kubernetes-reliability.md#overview) on this subject.

Reply:
how can we protect from following scenario:

1. customer install cluster with 3 workers nodes (default latency profile)
2. customer switch the node latency profile to LUSR (low update/slow reaction)
3. customer add 10000 worker nodes (edge?)
4. customer run for some time
5. customer switch the latency profile to "fast" for some reason
6. customer bricked their cluster, API server is under DDoS, and nodes are set as not ready

Reply:
The initial idea as described in https://issues.redhat.com/browse/WRKLDS-351 and https://issues.redhat.com/browse/OCPNODE-638 was only touching kube-controller-manager and kubelet. If you're planning to expand this to all control-plane components then this requires both @sttts and @hexfusion approval since they own api and etcd, respectively. Most importantly, though, it wasn't never announced to them that this will affect them so they couldn't account for that work in 4.10 nor in future versions.

Reply:
@soltysh It's correct that the initial idea was only touching kube-controller-manager and kubelet. However, as soon as we started writing this enhancement we realized that if you don't modify the relevant arguments of kube API server, it may have undesirable consequence on the pods that do no explicitly set taint based eviction timers. More on that in detail is described [here](https://github.com/kubernetes-sigs/kubespray/blob/master/docs/kubernetes-reliability.md#failure).

So in order to set the taint based eviction timers globally (and not limited to just the pods that explicitly set them), it's suggested that we can modify `--default-not-ready-toleration-seconds` & `--default-unreachable-toleration-seconds` arguments of the Kube API Server. This way we get a double protection by making sure the toleration values are set globally and since we are setting them, we can make sure they are set to their correct values. This could avoid unplanned cluster downtime for customers who might accidently (and manually) put wrong values (per pod). 

But I totally agree with your point that, since it was not announced to the API server team about this (in our defence, the moment we started looking into it we put those requirements as a part of this enhancement right away) so they may not able to account for this work in 4.10. But we can go ahead with just the changes in controller-manager and kubelet in 4.10 and cover in the documentation that customers must set taint based eviction timers manually (at least in 4.10). How does that sound @soltysh ?

Reply:
Although, since @sttts is suggesting we restrict the enhancement strictly to worker nodes, I think we might as well just drop the changes Kube API Server arguments which affect global toleration values, because I fear they will also affect pods running on the master nodes. 

In this case, we just have to restrict ourselves with the changes to controller manager and kubelet only but, make sure we document that user will have to set pod taint based eviction timers manually.

Reply:
The changes to kube-apiserver flags are fine. That's not what I meant. In other words, in my view there are two topics:

1. configure kube-apiserver (and kcm and kubelet) to cope with slowly and unreliably connected worker nodes
2. to cope with slow and unreliably connected master nodes.

I was arguing against (2). It must be clear in the enhancements that master nodes are connected to eachother as in any other non-edge scenario, with low latency and nearly 100% reliability. Only the workers are distant and hence we need (1).

Reply:
If there is any chance changing `--default-not-ready-toleration-seconds` & `--default-unreachable-toleration-seconds` args of the kube API server will affect any control plane pods then it's a strict no go. The more I think about it, more I am getting inclined to drop it. 

I had added the changes to Kube API server to minimize the error arising from user manually setting eviction taints for every pod, but that should not happen at the cost of destabilizing existing control plane pods. Let the user set those taints manually per pod, or as I have mentioned in some other comment we could explore adding an admission controller to inject those taints for all the pods heading for worker nodes.